apicast/src/oauth/keycloak.lua

```diff
@@ -0,0 +1,252 @@
+end
+
+function _M.new(config)
+  local configuration = config or _M.configuration
+
+  local is_valid = validate_config(configuration)
+
+  if not is_valid then
+    ngx.log(ngx.ERR,'Keycloak is not configured')
+    return error('missing keycloak configuration')
+  end
+
+  local http_client = http_ng.new{
+    backend = configuration.client,
+    options = {
+      ssl = { verify = false }
+    }
+  }
+
+  local keycloak_config = {
+    endpoint = configuration.endpoint,
+    authorize_url = resty_url.join(configuration.endpoint,'/protocol/openid-connect/auth'),
+    token_url = resty_url.join(configuration.endpoint,'/protocol/openid-connect/token'),
+    public_key = configuration.public_key or get_public_key(http_client, configuration.endpoint)
+  }
+
+  return setmetatable({
+    config = keycloak_config,
+    http_client = http_client
+    }, mt)
+end
+
+function _M.respond_and_exit(status, body, headers)
+  -- TODO: is there a better way to populate the response headers?..
+  if headers then
+    for name,value in pairs(headers) do
+      ngx.header[name] = value
+    end
+  end
+
+  ngx.status = status
+  ngx.print(body)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+function _M.respond_with_error(status, message)
+
+  --TODO: as per the RFC (https://tools.ietf.org/html/rfc6749#section-5.2) return WWW-Authenticate response header if 401
+  local headers = {
+    ['Content-Type'] = 'application/json;charset=UTF-8'
+  }
+  local err_msg = { error = message }
+  local body = cjson.encode(err_msg)
+  _M.respond_and_exit(status, body, headers)
+end
+
+function _M.authorize_check_params(params)
+  local response_type = params.response_type
+  local required_params = _M.params.response_type
+  if not response_type then return false, 'invalid_request' end
+  if not required_params[response_type] then return false, 'unsupported_response_type' end
+
+  for _,v in ipairs(required_params[response_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+
+  return true
+end
+
+function _M.token_check_params(params)
+  local grant_type = params.grant_type
+  local required_params = _M.params.grant_type
+  if not grant_type then return false, 'invalid_request' end
+  if not required_params[grant_type] then return false, 'unsupported_grant_type' end
+
+  for _,v in ipairs(required_params[grant_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+  return true
+end
+
+-- Parses the token - in this case we assume it's a JWT token
+-- Here we can extract authenticated user's claims or other information returned in the access_token
+-- or id_token by RH SSO
+function _M.parse_and_verify_token(jwt_token, public_key)
+  local jwt_obj = jwt:verify(public_key, jwt_token)
+  if not jwt_obj.verified then
+    ngx.log(ngx.INFO, "[jwt] failed verification for token: ", jwt_token)
+    return nil
+  end
+  return jwt_obj
+end
+
+function _M.check_credentials(params)
+  local http = {
```

Comment:
this does not need to be local variable of this function

it would be created on every call which is in the hot path and not JIT because it allocates new function